### ISSUE TITLE
FileAppenderCache/FileTarget issue

### ISSUE THREAD

**Issue by bjornbouetsmith** [NLog/NLog#1457]
**Type** Bug/Feature request

**NLog version**: 4.3.3

**Platform**: .Net 4.5

In case of a BUG:

The Current FileAppenderCache usage is kind of broken in my oppinion.

If you have two file targets in the same process, (KeepFileOpen=true, ConcurrentWrites=true, NetworkWrites=false, Archiving=true)trying to write to the same file, then they will each get their own instance of a FileAppenderCache, which then result in both FileTarget's getting their own instance of the RetryingMultiProcessFileAppender.

They both open a windows handle to the log file.
When archiving kicks in, they both start a thread to monitor the archive directory.

In this scenario and any other scenario when two different targets write to the same file within the same process, they should share the same instance of the Appender, which would cause one less thread per target to be created, and also only have one appender to have responsibility of the archiving, and also only one windows handle opened.

The current implementation allows for one target to use a SingleProcessFileAppender and another target to use a MutextMultiProcessFileAppender within the same process, which makes no sense at all, since one appender locks on a mutex to write to the file, the other just writes.

The reason why I think this should be changed is because I am being hit by somewhat the same issues as https://github.com/NLog/NLog/issues/1314
where the old old file is deleted, but another appender has a windows handle open on the file, and then the method: 
DeleteAndWaitForFileDelete never returns and keeps spinning in the 
do
                {
                    Thread.Sleep(100);
                    currentFileInfo = new FileInfo(fileName);
                }
                while ((currentFileInfo.Exists) && (currentFileInfo.CreationTime == originalFileCreationTime));

**Comment by bjornbouetsmith** [comment]
If I Change the line:
File.Delete(fileName) 

back to the original version:

  if (ConcurrentWrites && KeepFileOpen)
            {
                // Moving the file will cause all other FileTarget's to release the file, unlocking it for deleting.
                string tempFileName = Path.GetTempFileName();
                File.Delete(tempFileName);
                File.Move(fileName, tempFileName);
                File.Delete(tempFileName);
            }
            else
                File.Delete(fileName);

Then everything works as expected, meaning, file is deleted, and thread does not spin forever.

**Comment by bjornbouetsmith** [comment]
oh and if I have the debugger attached and steps through the code to the archiving, then everything works as expected, even when the unmodified code.

**Comment by 304NotModified** [comment]
Thanks, will looks at this.  

To be clear, code on previous versions was better?

**Comment by bjornbouetsmith** [comment]
Yes kind of, but I think fixing/enhancing the FileAppenderCache/FileAppender issue as described above will fix the issue as well, but that is a bigger task, so a quick fix to fix both this and https://github.com/NLog/NLog/issues/1314 would be to reintroduce the File.Move again as above.

**Comment by 304NotModified** [comment]
Fixing that issue would be nice also. If you have some thoughts on the code, please share it in a PR. :)

**Comment by bjornbouetsmith** [comment]
okay, I will make a PR with suggested code, I dont know if you can assign it to me, just so nobody else starts working on it?

**Comment by 304NotModified** [comment]
Cool! Thanks!

**Comment by bhaeussermann** [comment]
With regards to this (old) code:

```
// Moving the file will cause all other FileTarget's to release the file, unlocking it for deleting.
string tempFileName = Path.GetTempFileName();
File.Delete(tempFileName);
File.Move(fileName, tempFileName);
File.Delete(tempFileName);
```

The comment in that code is incorrect. When a file is moved to a different folder on the same disk, the OS only changes the FAT while the file itself remains untouched. It is still the same file and some file handles will still have a lock on that file and in fact still write to it even though it had been moved to a different folder.

**Comment by bhaeussermann** [comment]
Using two targets to the same file with different file appenders is not intended usage, and therefore the behavior for it is undefined.  
Perhaps we should mention this on the FileTarget page of the Wiki.

**Comment by bhaeussermann** [comment]
@bjornbouetsmith I'm curious. Why would you want to use two different targets to the same file?

**Comment by bjornbouetsmith** [comment]
Okay, I am not saying the comment is correct, but when I changed the new code back to the old code it started working again on my machine.
I would not use two different appenders on the same file ever. In fact I would never use two appenders on the same file, that is just silly in my opinion, but we have tests that does this and is was the tests that was failing for me.

But there are other issues in FileTarget as well in regards to access to the underlying file. Sometimes it goes through the appender to write to the file, other times it just opens a stream and writes directly.

Changing the code to use a single fileappender per file per process would prevent som of the locking/handle issues. And it would not make it worse for multiple processes writing to the same file.

But there is always a risk that something from another process opens a handle on the file being locked to, and then the archiving process would be stuck because of the loop.

do
{
Thread.Sleep(100);
currentFileInfo = new FileInfo(fileName);
}

So to prevent that it would probably be a better strategy to abort the loop after a configurable timeout and simply choose another filename for logging - to handle the handle locking issue.

I know that e.g. the WCF framework does this if you enable wcf diagnostics logging and it tries to cycle the file and cannot do it, then it simply creates a random name for the log file.
We dont have to do that, we can simply decide on another more fitting name.

**Comment by bjornbouetsmith** [comment]
But take a look at the stats in 
https://github.com/NLog/NLog/pull/1397
That shows that pooling can do for GC collection rates

**Comment by bhaeussermann** [comment]
> Changing the code to use a single fileappender per file per process would prevent som of the locking/handle issues.

Is this under the assumption that only one target is used per process?

**Comment by bjornbouetsmith** [comment]
No multiple targets, but each target referencing the same file should use the same appender, not have an appender each.

**Comment by bhaeussermann** [comment]
@bjornbouetsmith Users of NLog are not supposed to use multiple targets to the same file. Is it worth the additional complexity of the code to support this kind of unintended usage?

**Comment by bjornbouetsmith** [comment]
Thats true, but shouldn't we then throw an exception if someone sets up two targets for the same file? To prevent any threading/locking issue?
We have unit tests that does this, if we are not supposed to do it, then why are we testing it?

**Comment by 304NotModified** [comment]
I think throwing an exception beforehand will be a breaking change. 

About the unit test, have to search in git history as I dunno if was supported or a silly unit test.

**Comment by bjornbouetsmith** [comment]
Yes only throw exception if throwExceptions=true - otherwise we should ignore target number 2 - or even better don't setup anything. Isen't that what happens if you give nlog an invalid configuration?

**Comment by bhaeussermann** [comment]
There is a unit test that uses two `FileTarget` objects to the same file to simulate two processes logging to the same file.

**Comment by 304NotModified** [comment]
@bhaeussermann which unit test is that?

**Comment by 304NotModified** [comment]
@bhaeussermann any idea if I should delay the release of NLog 4.3.4 until this is fixed?

**Comment by bhaeussermann** [comment]
> which unit test is that?

It is `DateArchive_AllLoggersTransferToCurrentLogFile()`

**Comment by bhaeussermann** [comment]
> any idea if I should delay the release of NLog 4.3.4 until this is fixed?

The change in question was introduced in [#1183](https://github.com/NLog/NLog/pull/1183)

**Comment by 304NotModified** [comment]
> It is DateArchive_AllLoggersTransferToCurrentLogFile()

That one is recently added (7dae8cd57fc6ac50130085359d7912c684647a8a)  - jan 2016. 

> The change in question was introduced in #1183

So introduced in 4.3.0. I will wait with the release of 4.3.4 if we now better what to do with this issue. 

I think we can drop the support (in the unit tests) of writing 2 targets to the same file if this will help us fixing/improving the code. I don't prefer an exception, but writing a warning to the internal logger would be nice.

**Comment by bhaeussermann** [comment]
@304NotModified Based on the fact that we have never explicitly supported multiple targets to the same file within the same process, do you think that [#1183](https://github.com/NLog/NLog/pull/1183) actually broke something that was not broken before?

**Comment by bhaeussermann** [comment]
@304NotModified ('m not denying that this needs to be fixed; I'm just questioning whether the release should be delayed because of this).

**Comment by bjornbouetsmith** [comment]
I have to agree with @bhaeussermann - I dont think we should hold back any release. Its already out there, people using the "feature" are silly. But I think we should get it fixed, so that 
1. either you cannot have more than one target for one file per process
2. If you do, they share the fileappender cache, so that there is only one responsibility for archiving etc.

**Comment by 304NotModified** [comment]
OK thanks for the opinions! Will release 4.3.4 soon then   

Too bad I haven't found much time these days to dive into this issue.  Will try to be more informed this week.

**Comment by 304NotModified** [comment]
@snakefoot any idea if this is still an issue? We changed a lot with this.

**Comment by snakefoot** [comment]
> @304NotModified any idea if this is still an issue? 

The original issue was missing synchronization when having multiple isolated file-targets that are writing to the same file within the same application. Similar missing synchronization can also be found if having multiple network-targets pointing to the same network-address.

I'm guessing the problem was triggered by running unit-tests in parallel. Most of the old unit-test rely on a global logfactory, and some also fail to use guid-filename in temp-path. We have now disabled the ability to run unit-tests in parallel to avoid these issues.

Right now the application-owner have to adjust the config, when having multiple application-instances writing to the same file. I think the same can be expected from the application-owner, when having a single application-instance with multiple file-targets that might overlap in filenames. 

I think that the extra complexity needed to synchronize state between multiple Nlog-targets of the same type is not worth the price of code-complexity and performance.  So I would say wontfix.

**Comment by 304NotModified** [comment]
Thanks for the input :)
